### ISSUE TITLE
On making an AST

### ISSUE THREAD

**Issue by meisl** [irmen/prog8#52]
*[
Here's a little write-up of how I currently understand things to work in `compilerAst`, parts of which may be of use for a future guide on how to get going/engaged with development of the compiler.
In addition I'm trying to evaluate the pros and cons of the current approach and offer some ideas re possible other approaches.
<br>@irmen: this is NOT URGENT AT ALL, rather a recap for myself and some outlook. Up (maybe) only after "the plan" (CompilerDevelopment.md).
]*

---

## General organization (prep/basic, skip if you're already in the know)
* The Prog8 compiler is written in Kotlin. It uses ANTLR4 as its parser generator.
* The next best thing to Kotlin that ANTLR can generate is Java. All of that is under `parser/`.
* We don't want to write any Java ourselves, much less *tests* in Java.
* Hence there's `compilerAst/`, which is to provide an interface to this generated thing - purely in Kotlin. Here's also where we put tests of the grammar (via the Kotlin parser), and of the AST nodes.
* The parser in Java generated by ANTLR produces a *parse tree*, very closely reflecting the grammar.
* The other purpose of `compilerAst` is to turn this parse tree into an *AST* (Abstract Syntax Tree), which
  - consists only of (Kotlin) `prog8.parser.ast.Node`s, no more (Java) `prog8.parser.Prog8ANTLRParser.ParserRuleContext`s.
  - is somewhat more abstract, meaning: leaving out / summarizing unnecessary detail about the actual parse [TODO: example].
* The *unit of work*, as seen by the parser itself, is a *Prog8 module*: a piece of Prog8 source text which on the top level has *directives* and *blocks*, the latter of which then contain variable declarations, subroutines, ...
* The parser itself *does NOT do `%import` resolution*, it simply puts such directives (similar: `%asminclude`, `%asmbinary`) in the AST as is.

## Current approach
Aside from error handling and possibly recovery, and looking only at the parse tree -> AST conversion:

* Happens in `Antlr2Kotlin.kt`, **via extension methods** `.toAst(...)` on `ParserRuleContext` subclasses.
* Involved as well: ext. method `.toPosition()` (also in `Antlr2Kotlin.kt`).
* This works recursively, started by calling `ModuleContext.toPosition(..)`, which is the only such ext. method visible from the outside.
* Due to the (current) nature of almost all AST `Node` constructors, child nodes have to be instantiated *before* the parent node can be.

### Pros
* no subclassing of Java classes needed
* implicit `this` in ext. methods (here: the `ParserRuleContext` instance), hence no need to write it (or "it"...)
* (maybe) feels "natural", particularly the explicit recursion

### Cons
* *Because* of ext. methods, any global state about the current conversion can be shared *only* via parameter passing
* There *is* some such global state necessary:
  - `IStringEncoding` (atm, but hopefully soon to go, see #50)
  - the current *Prog8 module*, aka unit of work. Of that at least the *origin* (file system, embedded library/resources, plain String) is needed, particularly for equipping every `Node` with a proper `Position` (-> ext. method `.toPosition()`)
* Therefore: if we want to *change* what's shared global state, then that means going over *each and every ext. method's **declaration as well as invocation***, without much help from the IDE.
* Unfortunately, due to the recursive, bottom-up nature, the actual `Module` instance is built *at the very last*. Likewise for any other non-leaf `Node`...
* Similarly, every `Node`s `.parent` has to be `lateinit`; and there's `Node.linkParents(...)` (although this linking might be possible without it, even with the current approach)
* On the other hand, a `Node`s `.position` is (atm) NOT `lateinit`, leading to a variety of weird and intricate problems (the `Position`s `.file` being the cause).

### First, let's get some "givens" out of the way:

* I am assuming that we'll keep the *general* approach of not using the Java `ParserRuleContext` instances from the ANTLR-generated parser *for anything other than just turning them into our own Kotlin objects*.
Even though this means a good deal of simple copying-over, I, personally, am fine with that. Be it only for the unbearably ugly names: `XyzContext`..!
* By default, the generated parser collects all children (parse-tree!) nodes - ie. *including* `EOL` tokens, commas, etc. - in a `children` list on each `ParserRuleContext`, but that can be switched off. Atm, it is switched on, but we don't use any of that (I think). It's probably not so big an impact, but still work to do for the garbage collector.
* ANTLR optionally does provide a) a listener interface (= `enterXyzContext`/`exitXyzContext` "events") and/or b) a visitor interface (= `visitXyzContext(...)` callbacks). However, as alluded to in parentheses, neither of them uses *double-dispatch*, meaning you're stuck with those all-to-chatty names. Currently, we don't use them, and neither of these options are enabled.
* The generated parser *can* be made to do a lot more, only by editing the grammar. That goes from collecting results from subrules into lists (trivial code, way simpler than Java or Kotlin) up to semantic predicates and pre/post code blocks in the target language (= Java). The latter is definitely too much, but we could profit from the former. Currently, we're using practically zero.

### Finally, actual ideas:

---

*[
Remark: from here on it's pretty much "TODO"/rough sketch. Sorry for that, but it's late and a lot already. I'll edit and amend this OP later, possibly incorporating comments, if any.
]*

---

During conversion, it would be nice to have the sharing of state...

* ...*not* through parameters
* ...but rather simpler, as common closed-over scope (usually an "object" in OO, but also in the traditional sense)
* Again: this should make a lot - if not all - of those "when-to-initialize-what-but-dont-have-it-yet" problems go away (TODO: proof...)
* Both, the Listener and Visitor pattern provide for that. Of course in addition to - just like ext. methods - the ability to "attach on after the fact".
* Unfortunately neither of them are provided by ANTLR in an acceptable form, so we have to make them ourselves.
* Well, why not? Has been done before for Kotlin AST nodes: `IAstVisitor`, `AstWalker` (even though somewhat "unusual").
* Have actually tried out both of the below with Prog8Parser, so not purely theoretical.

#### Alternative 1: (self-made) parse tree **visitor** (client driven, "pull-model") 
* too much boiler-plate as it stands (even though once written, much less so for clients)
* benefit over listener would be that the client can actively steer the traversal
* but: that's actually not needed here (in entirety)
* and: additional burden on client code: must call super.visit(..)
* since the normal case *is* to visit the children, that means *more* client boiler-plate in the normal case

TODO: example client code

#### Alternative 2: (self-made) parse **listener** (event-driven, "push-model")
* still a lot repetitive code to be written, but about half compared to visitor (for the client: even less)
* plus: can use Kotlin's type system to help us when introducing or removing AST `Node` subtypes
* still possible to e.g. request to "skip children" from client
* simpler client interface: you don't do anything in the normal case (see above), but only when you do skip children
* easier to comprehend for readers new to the code base (if done right)

TODO: example client code

**Comment by irmen** [comment]
"a little writeup" :smile:

**Comment by meisl** [comment]
Oh well...

**Comment by irmen** [comment]
@meisl a little history:  I read a few articles by  Federico Tomassetti about antlr and kotlin ast's, I think one of them was https://tomassetti.me/parse-tree-abstract-syntax-tree/   and that put me on the way of the 'toAst()' extension methods. They've since stuck as the mechanism to get the kotlin ast classes from the antlr syntax tree.


Finally, there's this https://github.com/Strumenta/antlr-kotlin/   but it's still experimental .   It could perhaps be used to remove the antlr java parser code and replace it with kotlin parser code directly -- but we still want to convert the antlr syntax tree to our own specialized ast classes.

**Comment by irmen** [comment]
while the complex "compiler Ast" as discussed above is still generated from the antlr parser, there is now also a new lightweight "codeAst" that is meant to be used in the codegenerator (this still has to be built)
So the idea is that the  compilerAst is going to be internal to the compiler and optimizer phases and is no longer needed/visible in the codegenerators.

**Comment by irmen** [comment]
At least all code generator backends now are no longer dependent on the "compiler Ast" but instead use that simpler "code Ast" instead.  I'm finalizing this on the "codegen-on-new-ast" branch.  All code generators share the same interface and only see that new ast and the symbol table.

**Comment by irmen** [comment]
This has been an interesting collection of observations, but they're no longer relevant to the current state of how the compiler works.  

At least not for the "public" version of the Ast that is being used in the code generators.
(the internal compiler Ast is still mostly what it was, maybe it will be moved onto the new Ast one day too but that is way too much refactoring work for now)
